scanner.py

```diff
@@ -1714,21 +1715,69 @@ def _is_faulty_coordinate(self, x, y, add_offsets=False):
         # Streaming mode
 
     def _check_hardware(self, sample):
+        # Validate sample input
+        if "data" not in sample or "freq" not in sample:
+            raise ValueError("Sample must contain 'data' and 'freq' keys.")
+
+        # Initialize variables on the first call
+        if not hasattr(self, "freq_window"):
+            self.freq_window = deque(maxlen=50)  # Sliding window of the last 50 readings
+            self.min_threshold = None  # Minimum frequency threshold
+
+        # Add the current frequency to the sliding window
+        freq = sample["freq"]
+        self.freq_window.append(freq)
+
+        # Calculate statistics from the sliding window
+        if len(self.freq_window) > 1:
+            f_avg = np.mean(self.freq_window)
+            f_std = np.std(self.freq_window)
+            dynamic_threshold = f_avg + 3 * f_std  # 3-sigma threshold
+        else:
+            # Fallback during initialization
+            f_avg = freq
+            f_std = 0
+            dynamic_threshold = freq * 1.2  # Example: 20% above initial value
+
+        # Ensure a minimum threshold is set
+        if self.min_threshold is None:
+            self.min_threshold = freq * 1.2  # Set during the first function call
```

Comment:
Should we change these to not be magic numbers?
There was others I saw but cherry picked these. Not fussed but yeah.

Reply:
b410fd2 should fix this for you let me know if you dont like the way thats done